examples/ilu-preconditioned-solver-iterative-trisolves/ilu-preconditioned-solver-iterative-trisolves.cpp

```diff
@@ -0,0 +1,148 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+
+int main(int argc, char *argv[])
+{
+    // Some shortcuts
+    using ValueType = double;
+    using IndexType = int;
+
+    using vec = gko::matrix::Dense<ValueType>;
+    using mtx = gko::matrix::Csr<ValueType, IndexType>;
+    using gmres = gko::solver::Gmres<ValueType>;
+    using ir = gko::solver::Ir<ValueType>;
+    using bj = gko::preconditioner::Jacobi<ValueType, IndexType>;
+
+    // Print version information
+    std::cout << gko::version_info::get() << std::endl;
+
+    // Figure out where to run the code
+    std::shared_ptr<gko::Executor> exec;
+    if (argc == 1 || std::string(argv[1]) == "reference") {
+        exec = gko::ReferenceExecutor::create();
+    } else if (argc == 2 && std::string(argv[1]) == "omp") {
+        exec = gko::OmpExecutor::create();
+    } else if (argc == 2 && std::string(argv[1]) == "cuda" &&
+               gko::CudaExecutor::get_num_devices() > 0) {
+        exec = gko::CudaExecutor::create(0, gko::OmpExecutor::create());
+    } else if (argc == 2 && std::string(argv[1]) == "hip" &&
+               gko::HipExecutor::get_num_devices() > 0) {
+        exec = gko::HipExecutor::create(0, gko::OmpExecutor::create());
+    } else {
+        std::cerr << "Usage: " << argv[0] << " [executor]" << std::endl;
+        std::exit(-1);
+    }
+
+    // Read data
+    auto A = gko::share(gko::read<mtx>(std::ifstream("data/A.mtx"), exec));
+    auto b = gko::read<vec>(std::ifstream("data/b.mtx"), exec);
+    auto x = gko::read<vec>(std::ifstream("data/x0.mtx"), exec);
+
+    // Generate incomplete factors using ParILU
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<ValueType, IndexType>::build().on(exec);
+    // Generate concrete factorization for input matrix
+    auto par_ilu = par_ilu_fact->generate(A);
+
+    // Generate an iterative refinement factory to be used as triangular solver
+    // in the preconditioner application. The generated method is equivalent
+    // to doing five block-Jacobi sweeps with maximum block size 16.
+    auto bj_factory =
+        bj::build()
+            .with_max_block_size(16u)
+            .with_storage_optimization(gko::precision_reduction::autodetect())
+            .on(exec);
+
+    auto trisolve_factory =
+        ir::build()
+            .with_solver(share(bj_factory))
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(5u).on(exec))
```

Comment:
I think this is the sweeps parameter that you would like to tweak, right ? Can you make this settable by the user ?

Reply:
Do you mean something like this? Or should I make it optional and set a default?

Reply:
That should be good.